File: src/libcadet/model/binding/SaskaBinding.cpp
// =============================================================================
//  CADET - The Chromatography Analysis and Design Toolkit
//  
//  Copyright © 2008-2020: The CADET Authors
//            Please see the AUTHORS and CONTRIBUTORS file.
//  
//  All rights reserved. This program and the accompanying materials
//  are made available under the terms of the GNU Public License v3.0 (or, at
//  your option, any later version) which accompanies this distribution, and
//  is available at http://www.gnu.org/licenses/gpl.html
// =============================================================================

#include "model/binding/BindingModelBase.hpp"
#include "model/ExternalFunctionSupport.hpp"
#include "model/ModelUtils.hpp"
#include "cadet/Exceptions.hpp"
#include "model/Parameters.hpp"
#include "SlicedVector.hpp"
#include "LocalVector.hpp"
#include "SimulationTypes.hpp"

#include <functional>
#include <unordered_map>
#include <string>
#include <vector>

/*<codegen>
{
	"name": "SaskaParamHandler",
	"externalName": "ExtSaskaParamHandler",
	"parameters":
		[
			{ "type": "ScalarComponentDependentParameter", "varName": "h", "confName": "SASKA_H"},
			{ "type": "ComponentDependentComponentVectorParameter", "varName": "k", "confName": "SASKA_K"}
		]
}
</codegen>*/

/* Parameter description
 ------------------------
 h = Henry coefficient
 k = Quadratic factor
*/

namespace cadet
{

namespace model
{

inline const char* SaskaParamHandler::identifier() CADET_NOEXCEPT { return "SASKA"; }

inline bool SaskaParamHandler::validateConfig(unsigned int nComp, unsigned int const* nBoundStates)
{
	if ((_h.size() != _k.slices()) || (_k.size() != nComp * nComp) || (_h.size() < nComp))
		throw InvalidParameterException("SASKA_K has to have NCOMP^2 and SASKA_H NCOMP many elements");

	return true;
}

inline const char* ExtSaskaParamHandler::identifier() CADET_NOEXCEPT { return "EXT_SASKA"; }

inline bool ExtSaskaParamHandler::validateConfig(unsigned int nComp, unsigned int const* nBoundStates)
{
	if ((_h.size() != _k.slices()) || (_k.size() != nComp * nComp) || (_h.size() < nComp))
		throw InvalidParameterException("EXT_SASKA_K has to have NCOMP^2 and EXT_SASKA_H NCOMP many elements");

	return true;
}


/**
 * @brief Defines the Saska binding model
 * @details Implements the Saska adsorption model: \f[ \begin{align} 
 *              \frac{\mathrm{d} q_i}{\mathrm{d} t} = H_i c_{p,i} + \sum_{j=1}^{N_{\text{comp}}} k_{ij} c_{p,i} c_{p,j} - q_i,
 *          \end{align} \f]
 *          where @f$ H_i @f$ denotes the Henry coefficient and @f$ k_{ij} @f$ the quadratic factor.
 *          Multiple bound states are not supported. However, the quadratic factors use the bound phase parameter index for @f$ j @f$. 
 *          Components without bound state (i.e., non-binding components) are supported.
 *          
 *          See @cite Saska1992.
 * @tparam ParamHandler_t Type that can add support for external function dependence
 */
template <class ParamHandler_t>
class SaskaBindingBase : public ParamHandlerBindingModelBase<ParamHandler_t>
{
public:

	SaskaBindingBase() { }
	virtual ~SaskaBindingBase() CADET_NOEXCEPT { }

	static const char* identifier() { return ParamHandler_t::identifier(); }

	virtual bool implementsAnalyticJacobian() const CADET_NOEXCEPT { return true; }

	virtual void timeDerivativeQuasiStationaryFluxes(double t, unsigned int secIdx, const ColumnPosition& colPos, double const* yCp, double const* y, double* dResDt, LinearBufferAllocator workSpace) const
	{
		if (!this->hasQuasiStationaryReactions())
			return;

		if (!ParamHandler_t::dependsOnTime())
			return;

		// Update external function and compute time derivative of parameters
		typename ParamHandler_t::ParamsHandle p;
		typename ParamHandler_t::ParamsHandle dpDt;
		std::tie(p, dpDt) = _paramHandler.updateTimeDerivative(t, secIdx, colPos, _nComp, _nBoundStates, workSpace);

		// Protein fluxes: -H_i c_{p,i} - \sum_j k_{ij} c_{p,i} c_{p,j} + q_i
		unsigned int bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			const active h = dpDt->h[i];
			active const* const kSlice = dpDt->k[i];

			// Residual
			dResDt[bndIdx] = -static_cast<double>(h) * yCp[i];
			for (int j = 0; j < _nComp; ++j)
				dResDt[bndIdx] -= static_cast<double>(kSlice[j]) * yCp[i] * yCp[j];

			// Next bound component
			++bndIdx;
		}
	}

	CADET_BINDINGMODELBASE_BOILERPLATE

protected:
	using ParamHandlerBindingModelBase<ParamHandler_t>::_paramHandler;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_reactionQuasistationarity;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_nComp;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_nBoundStates;

	template <typename StateType, typename CpStateType, typename ResidualType, typename ParamType>
	int fluxImpl(double t, unsigned int secIdx, const ColumnPosition& colPos, StateType const* y,
		CpStateType const* yCp, ResidualType* res, LinearBufferAllocator workSpace) const
	{
		typename ParamHandler_t::ParamsHandle const p = _paramHandler.update(t, secIdx, colPos, _nComp, _nBoundStates, workSpace);

		// Protein fluxes: -H_i c_{p,i} - \sum_j k_{ij} c_{p,i} c_{p,j} + q_i
		unsigned int bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			const active h = p->h[i];
			active const* const kSlice = p->k[i];

			// Residual
			res[bndIdx] = -static_cast<ParamType>(h) * yCp[i] + y[bndIdx];
			for (int j = 0; j < _nComp; ++j)
				res[bndIdx] -= static_cast<ParamType>(kSlice[j]) * yCp[i] * yCp[j];

			// Next bound component
			++bndIdx;
		}

		return 0;
	}

	template <typename RowIterator>
	void jacobianImpl(double t, unsigned int secIdx, const ColumnPosition& colPos, double const* y, double const* yCp, int offsetCp, RowIterator jac, LinearBufferAllocator workSpace) const
	{
		typename ParamHandler_t::ParamsHandle const p = _paramHandler.update(t, secIdx, colPos, _nComp, _nBoundStates, workSpace);

		// Protein fluxes: -H_i c_{p,i} - \sum_j k_{ij} c_{p,i} c_{p,j} + q_i
		int bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			const double h = static_cast<double>(p->h[i]);
			active const* const kSlice = p->k[i];

			// dres_i / dc_{p,j}
			for (int j = 0; j < _nComp; ++j)
				jac[j - bndIdx - offsetCp] = -static_cast<double>(kSlice[j]) * yCp[i];
				// Getting to c_{p,i}: -bndIdx takes us to q_0, another -offsetCp to c_{p,0} and a +j to c_{p,j}.
				//                     This means jac[j - bndIdx - offsetCp] corresponds to c_{p,j}.

			// dres_i / dc_{p,i}
			for (int j = 0; j < _nComp; ++j)
				jac[i - bndIdx - offsetCp] -= static_cast<double>(kSlice[j]) * yCp[j];
			
			jac[i - bndIdx - offsetCp] -= h;

			// dres_i / dq_i
			jac[0] = 1.0;

			// Advance to next flux and Jacobian row
			++bndIdx;
			++jac;
		}
	}
};


typedef SaskaBindingBase<SaskaParamHandler> SaskaBinding;
typedef SaskaBindingBase<ExtSaskaParamHandler> ExternalSaskaBinding;

namespace binding
{
	void registerSaskaModel(std::unordered_map<std::string, std::function<model::IBindingModel*()>>& bindings)
	{
		bindings[SaskaBinding::identifier()] = []() { return new SaskaBinding(); };
		bindings[ExternalSaskaBinding::identifier()] = []() { return new ExternalSaskaBinding(); };
	}
}  // namespace binding

}  // namespace model

}  // namespace cadet

File: src/cadet-mi/bindings/SaskaBinding.m

classdef SaskaBinding < KineticQuasiStationaryBindingModel
	%SaskaBinding Saska binding model
	%
	% See also BINDINGMODEL, KINETICQUASISTATIONARYBINDINGMODEL

	% Copyright: (C) 2008-2024 The CADET Authors
	%            See the license note at the end of the file.

	properties(Constant)
		name = 'SASKA'; % Name of the binding model according to CADET file format specs
	end

	properties (Dependent, Transient)
		% Henri coefficient in [m^3_MP / (m^3_SP * s)]
		h;
		SASKA_H;
		% Quadratic factors in [m^6_MP / (m^3_SP * mol * s)]
		k;
		SASKA_K;
	end

	methods

		function obj = SaskaBinding(h, k)
			%SASKABINDING Constructs a SaskaBinding object
			%   OBJ = SASKABINDING(H) creates a SaskaBinding model with the
			%   given Henri coefficients H.
			%
			%   OBJ = SASKABINDING(..., K) also sets the quadratic factors to K.

			obj = obj@KineticQuasiStationaryBindingModel();

			if nargin >= 1
				obj.h = h;
			end
			if nargin >= 2
				obj.k = k;
			end
		end

		function res = validate(obj, nComponents, nBoundStates)
			%VALIDATE Validates the binding model parameters
			%   RES = VALIDATE(NCOMPONENTS, NBOUNDSTATES) uses the number of
			%   components NCOMPONENTS of the model and the number of bound
			%   states of each of these components as given by the vector
			%   NBOUNDSTATES to validate the parameters of the binding model.
			%   Returns true in RES if everything is fine and false otherwise.

			validateattributes(obj.h, {'double'}, {'vector', 'nonempty', 'finite', 'real', 'numel', nComponents}, '', 'h');
			validateattributes(obj.k, {'double'}, {'2d', 'nonempty', 'finite', 'real', 'size', [nComponents, nComponents]}, '', 'k');
			res = obj.validate@KineticQuasiStationaryBindingModel(nComponents, nBoundStates);
		end

		function val = get.h(obj)
			val = obj.data.SASKA_H;
		end

		function set.h(obj, val)
			validateattributes(val, {'double'}, {'vector', 'nonempty', 'finite', 'real'}, '', 'h');
			obj.data.SASKA_H = val;
			obj.hasChanged = true;
		end

		function val = get.k(obj)
			val = obj.data.SASKA_K;
			val = reshape(val, sqrt(numel(val)), sqrt(numel(val))).';
		end

		function set.k(obj, val)
			validateattributes(val, {'double'}, {'2d', 'nonempty', 'finite', 'real'}, '', 'k');
			nVals = sqrt(numel(val));
			if nVals ~= ceil(nVals)
				error('CADET:invalidConfig', 'Expected k to have a square number of values.');
			end
			val = val.';
			obj.data.SASKA_K = val(:);
			obj.hasChanged = true;
		end

		function val = get.SASKA_H(obj)
			val = obj.h;
		end
		function set.SASKA_H(obj, val)
			obj.h = val;
		end
		function val = get.SASKA_K(obj)
			val = obj.k;
		end
		function set.SASKA_K(obj, val)
			obj.k = val;
		end
	end

	methods (Static)

		function obj = loadobj(S)
			if isstruct(S)
				obj = SaskaBinding();
				obj.loadobjInternal(S);
			end
		end
		
	end

end

% =============================================================================
%  CADET
%  
%  Copyright (C) 2008-2024: The CADET Authors
%            Please see the AUTHORS and CONTRIBUTORS file.
%  
%  All rights reserved. obj program and the accompanying materials
%  are made available under the terms of the GNU Public License v3.0 (or, at
%  your option, any later version) which accompanies obj distribution, and
%  is available at http://www.gnu.org/licenses/gpl.html
% =============================================================================
